{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "This code makes a FHIR Patient search request to retrieve a patient record using an NHS number identifier. Hereâ€™s a clear, structured description of what the request is doing.\n",
    "\n",
    "High-level description\n",
    "\n",
    "The request performs an HTTP GET against a FHIR serverâ€™s Patient endpoint, searching for a patient whose identifier matches a given NHS number.\n",
    "Authentication is handled using HTTP Basic Auth, with credentials loaded from environment variables.\n",
    "\n",
    "Breakdown of the FHIR request\n",
    "1. Endpoint being called\n",
    "GET {FHIR_SERVER}/Patient?identifier=https://fhir.nhs.uk/Id/nhs-number|9999999522\n",
    "\n",
    "\n",
    "Resource type: Patient\n",
    "\n",
    "Operation: Search\n",
    "\n",
    "Search parameter: identifier\n",
    "\n",
    "2. Search parameter details\n",
    "identifier=https://fhir.nhs.uk/Id/nhs-number|9999999522\n",
    "\n",
    "\n",
    "This follows the FHIR search syntax for identifiers:\n",
    "\n",
    "identifier = system | value\n",
    "\n",
    "\n",
    "system: https://fhir.nhs.uk/Id/nhs-number\n",
    "Identifies the namespace for NHS numbers (UK-specific FHIR identifier system)\n",
    "\n",
    "value: 9999999522\n",
    "The NHS number being searched for\n",
    "\n",
    "ðŸ‘‰ This tells the FHIR server:\n",
    "\n",
    "â€œReturn any Patient resource that has an identifier with this system and value.â€"
   ],
   "id": "4cf27ec24263825d"
  },
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2026-01-07T06:32:18.204073Z",
     "start_time": "2026-01-07T06:32:18.134053Z"
    }
   },
   "source": [
    "import requests\n",
    "import fhirclient.models.patient as pat\n",
    "from requests.auth import HTTPBasicAuth\n",
    "from dotenv import load_dotenv\n",
    "load_dotenv()\n",
    "import os\n",
    "\n",
    "fhir_password = os.getenv(\"FHIR_PASSWORD\")\n",
    "fhir_username = os.getenv(\"FHIR_USERNAME\")\n",
    "server = os.getenv(\"FHIR_SERVER\")\n",
    "\n",
    "nhsNumber = os.getenv(\"NHS_NUMBER\")\n",
    "if nhsNumber == None:\n",
    "    nhsNumber = '9999999522'\n",
    "\n",
    "api_url = server + \"Patient?identifier=https://fhir.nhs.uk/Id/nhs-number|\"+nhsNumber\n",
    "print(\"Calling GET \"+api_url)\n",
    "response = requests.get(api_url, auth=HTTPBasicAuth(fhir_username, fhir_password))\n",
    "patientJSON = response.json()\n",
    "#print(patientJSON)\n",
    "\n",
    "patientId = None\n",
    "\n",
    "if ((patientJSON['total']> 0) and (len(patientJSON['entry'])>0)):\n",
    "    patientId = patientJSON['entry'][0]['resource']['id']\n",
    "    patient = pat.Patient(patientJSON['entry'][0]['resource'])\n",
    "    print(\"Patient Id: \" + patientId)\n",
    "    print(\"Date of Birth: \"+patient.birthDate.isostring)\n",
    "    for id in patient.identifier:\n",
    "        print('Identifier: '+ id.value + \" \" + id.type.coding[0].code)\n",
    "\n"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calling GET http://192.168.1.67/healthconnect/cdr/fhir/r4/Patient?identifier=https://fhir.nhs.uk/Id/nhs-number|9999999522\n",
      "Patient Id: 96\n",
      "Date of Birth: 1945-01-01\n",
      "Identifier: 9999999522 NH\n",
      "Identifier: 20002983 MR\n"
     ]
    }
   ],
   "execution_count": 10
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "",
   "id": "ceec0e3352760c26"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-07T06:32:18.469479Z",
     "start_time": "2026-01-07T06:32:18.219548Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import fhirclient.models.diagnosticreport as dr\n",
    "import fhirclient.models.meta as meta\n",
    "import pandas as pd\n",
    "from dateutil import parser\n",
    "\n",
    "def issued(issued):\n",
    "    if issued == None:\n",
    "        return None\n",
    "    return parser.parse(issued.isostring)\n",
    "\n",
    "def lastUpdated(meta : meta.Meta):\n",
    "    if meta == None:\n",
    "        return None\n",
    "    return parser.parse(meta.lastUpdated.isostring)\n",
    "\n",
    "if patientId != None:\n",
    "    api_url = server + \"DiagnosticReport?patient=\"+patientId+\"&_sort=-issued\"\n",
    "    response = requests.get(api_url, auth=HTTPBasicAuth(fhir_username, fhir_password))\n",
    "    response1JSON = response.json()\n",
    "\n",
    "    print(response1JSON['total'])\n",
    "\n",
    "    diagnosticReports = []\n",
    "    for entry in response1JSON['entry']:\n",
    "        #print(entry['resource'])\n",
    "        print(entry['resource']['resourceType'], entry['resource']['issued'] )\n",
    "        report = dr.DiagnosticReport(entry['resource'])\n",
    "        diagnosticReports.append(report)\n",
    "\n",
    "    dfDR = pd.DataFrame([vars(s) for s in diagnosticReports])\n",
    "    dfDR['issued'] = dfDR['issued'].apply(issued)\n",
    "    dfDR['effectiveDateTime'] = dfDR['effectiveDateTime'].apply(issued)\n",
    "    dfDR['lastUpdated'] = dfDR['meta'].apply(lastUpdated)\n",
    "\n",
    "dfDR"
   ],
   "id": "bde845135209d277",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "DiagnosticReport 2025-12-06T13:37:35+00:00\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "                                             basedOn  \\\n",
       "0  [<fhirclient.models.fhirreference.FHIRReferenc...   \n",
       "\n",
       "                                            category  \\\n",
       "0  [<fhirclient.models.codeableconcept.CodeableCo...   \n",
       "\n",
       "                                                code conclusion  \\\n",
       "0  <fhirclient.models.codeableconcept.CodeableCon...       None   \n",
       "\n",
       "  conclusionCode         effectiveDateTime effectivePeriod encounter  \\\n",
       "0           None 2025-12-06 13:37:35+00:00            None      None   \n",
       "\n",
       "                                          identifier imagingStudy  ...  \\\n",
       "0  [<fhirclient.models.identifier.Identifier obje...         None  ...   \n",
       "\n",
       "  modifierExtension  text  id implicitRules language  \\\n",
       "0              None  None  99          None     None   \n",
       "\n",
       "                                                meta _server _resolved _owner  \\\n",
       "0  <fhirclient.models.meta.Meta object at 0x108a2...    None      None   None   \n",
       "\n",
       "                lastUpdated  \n",
       "0 2026-01-07 04:25:42+00:00  \n",
       "\n",
       "[1 rows x 31 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>basedOn</th>\n",
       "      <th>category</th>\n",
       "      <th>code</th>\n",
       "      <th>conclusion</th>\n",
       "      <th>conclusionCode</th>\n",
       "      <th>effectiveDateTime</th>\n",
       "      <th>effectivePeriod</th>\n",
       "      <th>encounter</th>\n",
       "      <th>identifier</th>\n",
       "      <th>imagingStudy</th>\n",
       "      <th>...</th>\n",
       "      <th>modifierExtension</th>\n",
       "      <th>text</th>\n",
       "      <th>id</th>\n",
       "      <th>implicitRules</th>\n",
       "      <th>language</th>\n",
       "      <th>meta</th>\n",
       "      <th>_server</th>\n",
       "      <th>_resolved</th>\n",
       "      <th>_owner</th>\n",
       "      <th>lastUpdated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[&lt;fhirclient.models.fhirreference.FHIRReferenc...</td>\n",
       "      <td>[&lt;fhirclient.models.codeableconcept.CodeableCo...</td>\n",
       "      <td>&lt;fhirclient.models.codeableconcept.CodeableCon...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2025-12-06 13:37:35+00:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[&lt;fhirclient.models.identifier.Identifier obje...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>99</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>&lt;fhirclient.models.meta.Meta object at 0x108a2...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2026-01-07 04:25:42+00:00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 31 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 11
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-07T06:32:18.691549Z",
     "start_time": "2026-01-07T06:32:18.535381Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import fhirclient.models.documentreference as doc\n",
    "import pandas as pd\n",
    "from dateutil import parser\n",
    "\n",
    "\n",
    "if patientId != None:\n",
    "    api_url = server + \"DocumentReference?patient=\"+patientId\n",
    "    response = requests.get(api_url, auth=HTTPBasicAuth(fhir_username, fhir_password))\n",
    "    response1JSON = response.json()\n",
    "\n",
    "    print(response1JSON['total'])\n",
    "\n",
    "    documents = []\n",
    "    for entry in response1JSON['entry']:\n",
    "        document = doc.DocumentReference(entry['resource'])\n",
    "        documents.append(document)\n",
    "\n",
    "    dfDOC = pd.DataFrame([vars(s) for s in documents])\n",
    "    dfDOC['date'] = dfDOC['date'].apply(issued)\n",
    "    dfDOC['lastUpdated'] = dfDOC['meta'].apply(lastUpdated)\n",
    "    #dfDOC['effectiveDateTime'] = dfDR['effectiveDateTime'].apply(issued)\n",
    "\n",
    "dfDOC"
   ],
   "id": "859820c8ddbb2152",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "  authenticator author category  \\\n",
       "0          None   None     None   \n",
       "\n",
       "                                             content  \\\n",
       "0  [<fhirclient.models.documentreference.Document...   \n",
       "\n",
       "                                             context  \\\n",
       "0  <fhirclient.models.documentreference.DocumentR...   \n",
       "\n",
       "                                           custodian  \\\n",
       "0  <fhirclient.models.fhirreference.FHIRReference...   \n",
       "\n",
       "                       date description docStatus  \\\n",
       "0 2025-12-06 13:37:35+00:00        None      None   \n",
       "\n",
       "                                          identifier  ... modifierExtension  \\\n",
       "0  [<fhirclient.models.identifier.Identifier obje...  ...              None   \n",
       "\n",
       "   text   id implicitRules language  \\\n",
       "0  None  100          None     None   \n",
       "\n",
       "                                                meta _server _resolved _owner  \\\n",
       "0  <fhirclient.models.meta.Meta object at 0x10de7...    None      None   None   \n",
       "\n",
       "                lastUpdated  \n",
       "0 2026-01-07 04:25:42+00:00  \n",
       "\n",
       "[1 rows x 28 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>authenticator</th>\n",
       "      <th>author</th>\n",
       "      <th>category</th>\n",
       "      <th>content</th>\n",
       "      <th>context</th>\n",
       "      <th>custodian</th>\n",
       "      <th>date</th>\n",
       "      <th>description</th>\n",
       "      <th>docStatus</th>\n",
       "      <th>identifier</th>\n",
       "      <th>...</th>\n",
       "      <th>modifierExtension</th>\n",
       "      <th>text</th>\n",
       "      <th>id</th>\n",
       "      <th>implicitRules</th>\n",
       "      <th>language</th>\n",
       "      <th>meta</th>\n",
       "      <th>_server</th>\n",
       "      <th>_resolved</th>\n",
       "      <th>_owner</th>\n",
       "      <th>lastUpdated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[&lt;fhirclient.models.documentreference.Document...</td>\n",
       "      <td>&lt;fhirclient.models.documentreference.DocumentR...</td>\n",
       "      <td>&lt;fhirclient.models.fhirreference.FHIRReference...</td>\n",
       "      <td>2025-12-06 13:37:35+00:00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[&lt;fhirclient.models.identifier.Identifier obje...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>100</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>&lt;fhirclient.models.meta.Meta object at 0x10de7...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2026-01-07 04:25:42+00:00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 28 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 12
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-07T06:32:19.166834Z",
     "start_time": "2026-01-07T06:32:19.150944Z"
    }
   },
   "cell_type": "code",
   "source": "",
   "id": "2494621f6682cfbb",
   "outputs": [],
   "execution_count": 12
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-07T06:32:24.379075Z",
     "start_time": "2026-01-07T06:32:19.555250Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import base64\n",
    "import io\n",
    "import tabula\n",
    "from pypdf import PdfReader\n",
    "\n",
    "for document in documents:\n",
    "    content = document.content[0]\n",
    "    #print(content.attachment.contentType)\n",
    "    #print(content.attachment.url)\n",
    "    api_url = server + content.attachment.url\n",
    "    response = requests.get(api_url, auth=HTTPBasicAuth(fhir_username, fhir_password))\n",
    "    response1JSON = response.json()\n",
    "    encoded = response1JSON['data']\n",
    "    #print(encoded)\n",
    "    decode = base64.b64decode(encoded)\n",
    "    pdf_stream = io.BytesIO(decode)\n",
    "    pdf = PdfReader(pdf_stream)\n",
    "    print('Pages: '+ str(len(pdf.pages)))\n",
    "    print('============ PDF Reader ==============')\n",
    "    for page in pdf.pages:\n",
    "        print(page.extract_text())\n",
    "\n",
    "    tables = tabula.read_pdf(pdf_stream\n",
    "                             , pages='all', multiple_tables=True)\n",
    "    print('============ Tabula ==============')\n",
    "    for table in tables:\n",
    "        print(table)\n"
   ],
   "id": "36d1b2cefdd12c4a",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Failed to import jpype dependencies. Fallback to subprocess.\n",
      "No module named 'jpype'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pages: 2\n",
      "============ PDF Reader ==============\n",
      "North West Genomic Laboratory Hub (Manchester Site)\n",
      "Manchester Centre for Genomic Medicine\n",
      "6th Floor, St Mary's Hospital, Manchester, M13 9WL\n",
      "Scientific Operational Director: Dr E. Howard\n",
      "https://mft.nhs.uk/nwglh/  mft.genomics@nhs.net  Tel +44(0) 161 276 6122\n",
      " \n",
      "SevenForename:\n",
      "EDITTESTPATIENTSurname:\n",
      "Sex:\n",
      "01/01/1945 Sample:DoB: Blood\n",
      "R25-00HN-1Report No:\n",
      "Hospital No: Not provided\n",
      "Other No: Not provided\n",
      "NHS No: 999 999 9522\n",
      "Collected: Not provided Received: 21/11/2025 15:22 Activated: 21/11/2025\n",
      "Reported: 06/12/2025\n",
      "Referred by: Doctor Liverpool, Liverpool Womens Hospital, Liverpool, L8 7SS\n",
      "Path. No: ACC1973A1\n",
      "Genomics Laboratory Report\n",
      "Reason for Testing\n",
      "RESULT SUMMARY:\n",
      "Likely pathogenic variant detected.\n",
      "Likely genetic diagnosis of autosomal dominant XXXX OR GENE-related disease.\n",
      "This individual is heterozygous for a likely pathogenic GENE variant, VARIANT (details in Appendix II overleaf). Monoallelic pathogenic\n",
      "GENE variants cause (a range of disorders including) autosomal dominant XXXX (OMIM# XXXX).\n",
      "This result is likely to be the genetic cause of this individual's clinical symptoms, and would be consistent with an autosomal dominant mode\n",
      "of inheritance.\n",
      "[ONLY INCLUDE AS APPROPRIATE]: Testing of this individualâ€™s parents is available (via referral to a clinical genetics service) to\n",
      "investigate if the variant has arisen de novo, and to provide advice to their parents regarding future pregnancies, if appropriate.\n",
      "OR\n",
      "This result has implications for other family members. Testing for this variant is available to other relatives of this individual, as appropriate\n",
      "(via referral to a clinical genetics service).\n",
      "[REASON FOR TESTING - COPY BELOW TEXT TO 'REPORTING REFERRAL INFORMATION TEXT' BOX THEN DELETE]\n",
      "[Insert referring clinician notes]\n",
      "Clinical Indication: [Insert Clinical Indication Name (RXXX)].\n",
      "Result and Interpretation\n",
      "Jonathan Edgerley\n",
      "Clinical Scientist\n",
      " \n",
      "Authorised By:Reported By:\n",
      "Fraser Suttie\n",
      "Administrator\n",
      "Page 1 of 2\n",
      "SevenForename:\n",
      "EDITTESTPATIENTSurname: 01/01/1945DoB:\n",
      "Report No: R25-00HN-1\n",
      "APPENDIX I - Test Methodology:\n",
      "Clinical indication: Factor II deficiency (R112.1)\n",
      "Laboratory process\n",
      "DNA extraction: Chemagic Prime for samples extracted by NWGLH\n",
      "Enrichment method: Agilent SureSelect custom design\n",
      "Sequencing method: Illumina NextSeq 550\n",
      "Bioinformatics and quality control: Congenica â€“ Alignment and SNV/Indel calling â€“ Sentieon DNASeq; CNV calling â€“ Exome Depth\n",
      "Panel(s) applied: Factor II deficiency (R112.1)\n",
      "The panel includes diagnostic grade (green) genes from the associated PanelApp panel (https://nhsgms-panelapp.genomicsengland.co.uk/) for the specific version detailed\n",
      "above. This test is used to detect single nucleotide variants (SNVs), small insertions and deletions (INDELs) and copy number variations (defined as single exon or larger\n",
      "deletions and duplications). This panel targets protein coding exons, exon intron boundaries (Â± 8 bps) and selected non-coding, deep intronic variants. Only variants relevant\n",
      "to this patientâ€™s submitted phenotype have been reported. If the result does not fully account for the phenotype observed in the patient, please contact the laboratory to\n",
      "discuss if further analysis may be appropriate.\n",
      "Coverage: >97% of the coding regions of the diagnostic grade (green) genes in the panel above are covered to a minimum depth of 50X. Patient specific coverage for\n",
      "individual genes can be provided on request.\n",
      "Limitations: This test cannot be used for the detection of repeat expansions or variants in mitochondrial DNA (mtDNA). The test does not identify balanced translocations or\n",
      "complex inversions, and it may not detect low-level mosaicism or inter-locus gene conversion. The sensitivity to detect variants may be limited in genes where some, or all, of\n",
      "the gene is duplicated in the genome or the gene has suboptimal coverage. Please contact the laboratory for additional details.\n",
      "Confirmation of findings: SNV and INDEL calls which do not meet internally validated quality metrics are confirmed by Sanger sequencing. The presence, but not the\n",
      "extent, of CNV calls is confirmed by an independent method (MLPA, custom droplet digital PCR assay, or custom QF-PCR assay).\n",
      "Analytical validation: This laboratory-developed test has been locally validated. The sensitivity of this methodology for SNVs is 100% (95% CI: 99.78% to 100%), INDELs is\n",
      "95.86% (95% CI: 91.65% to 98.32%) and CNVs is 100% (95% CI: 93.28% to 100%), and specificity is >99.9% for most variant types. A normal result does not rule out the\n",
      "diagnosis of a genetic disorder since some DNA abnormalities are undetectable by the applied technology.\n",
      "Regulation and accreditations: This test is not UKAS accredited. This test has been validated locally and will be submitted for UKAS assessment imminently.\n",
      "Please contact the laboratory for additional details, if required.\n",
      "Page 2 of 2\n",
      "============ Tabula ==============\n",
      "    Forename:            Seven    Report No:        R25-00HN-1  Unnamed: 0  \\\n",
      "0    Surname:  EDITTESTPATIENT  Hospital No:      Not provided         NaN   \n",
      "1        Sex:              NaN     Other No:      Not provided         NaN   \n",
      "2        DoB:       01/01/1945       Sample:             Blood         NaN   \n",
      "3     NHS No:     999 999 9522           NaN               NaN         NaN   \n",
      "4         NaN              NaN     Path. No:         ACC1973A1         NaN   \n",
      "5  Collected:     Not provided     Received:  21/11/2025 15:22  Activated:   \n",
      "\n",
      "   Unnamed: 1  \n",
      "0         NaN  \n",
      "1         NaN  \n",
      "2         NaN  \n",
      "3         NaN  \n",
      "4         NaN  \n",
      "5  21/11/2025  \n"
     ]
    }
   ],
   "execution_count": 13
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
